runtime/common/src/main/kotlin/kotlinx/serialization/json/internal/TreeJsonInput.kt

```diff
@@ -0,0 +1,183 @@
+@file:Suppress("LeakingThis")
+
+package kotlinx.serialization.json.internal
+
+import kotlinx.serialization.*
+import kotlinx.serialization.internal.*
+import kotlinx.serialization.json.*
+
+internal fun <T> Json.readJson(element: JsonElement, deserializer: DeserializationStrategy<T>): T {
+    val descriptor = deserializer.descriptor
+    if (element is JsonNull) {
+        // TODO temporary workaround (?)
+        require(descriptor.isNullable) { "Read JsonNull and expected nullable descriptor, but has $descriptor" }
+        @Suppress("NULL_FOR_NONNULL_TYPE")
+        return null
+    }
+
+    val input = when (descriptor.kind) {
+        StructureKind.LIST -> JsonTreeListInput(this, cast(element))
+        StructureKind.MAP -> JsonTreeMapInput(this, cast(element))
+        is PrimitiveKind -> JsonPrimitiveInput(this, cast(element))
+        else -> JsonTreeInput(this, cast(element))
+    }
+
+    return input.decode(deserializer)
+}
+
+private sealed class AbstractJsonTreeInput(override val json: Json, open val obj: JsonElement)
+    : NamedValueDecoder(), JsonInput {
+
+    init {
+        context = json.context
+    }
+
+    override val updateMode: UpdateMode
+        get() = json.updateMode
+
+    override fun readTree(): JsonElement {
```

Comment:
It looks to me like this function should always either `return obj` or be equivalent to `currentElement()`